cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -114,6 +114,32 @@ abstract class Configuration extends string {
    */
   predicate hasFlowToExpr(DataFlowExpr sink) { hasFlowTo(exprNode(sink)) }
 
+  /**
+   * Gets the exploration limit for `hasPartialFlow` measured in approximate
+   * number of interprocedural steps.
+   */
+  int explorationLimit() { none() }
+
+  /**
+   * Holds if there is a partial data flow path from `source` to `node`. The
+   * approximate distance between `node` and the closest source is `dist` and
+   * is restricted to be less than or equal to `explorationLimit()`. This
+   * predicate completely disregards sink definitions.
+   *
+   * This predicate is intended for dataflow exploration and debugging and may
+   * perform poorly if the number of sources is too big and/or the exploration
+   * limit is set too high without using barriers.
+   *
+   * This predicate is disabled (has no results) by default. Override
+   * `explorationLimit()` with a suitable number to enable this predicate.
+   *
+   * To use this in a `path-problem` query, import the module `PartialPathGraph`.
+   */
+  final predicate hasPartialFlow(PartialPathNode source, PartialPathNode node, int dist) {
```

Comment:
For usability, would it be viable to make `source` be a plain `DataFlow::Node`, and make `node` be the `result`? That way you could use this in a query without needing to mention the new `PartialPathNode` type, making it easier to debug an already-written query. (One reason not to do this would be if you foresee the need to call `hasPartialFlow` with things that are not in `Configuration::isSource`.)

Reply:
Sure.  We could just have both options, i.e. add
```
final PartialPathNode getPartialFlow(Node source) {
  exists(PartialPathNode src | src.getNode() = source | this.hasPartialFlow(src, result, _))
}
```
How does that look?

Reply:
Or should we keep the `dist` parameter for easier discoverability? (It's also available directly from `PartialPathNode`)

Reply:
Keeping the dist parameter sounds better -- it only saves typing `, _` to omit it, and it's not like anyone will want to apply a TC to this predicate and hence need it to be binary.

Reply:
Hmm, before making this change: You need a `PartialPathNode` for both the source and the reached node, if you want to see the path - so in that use-case we need the `hasPartialFlow` predicate as-is.  With this in mind, would you still like a `getPartialFlow` predicate alongside the current `hasPartialFlow`?